source/game.production.script.bmx

```diff
@@ -137,51 +137,36 @@ Type TScriptCollection Extends TGameObjectCollection
 		Local template:TScriptTemplate
 		Local filter:SScriptTemplateFilter = new SScriptTemplateFilter
 		filter.biggerStudioAllowedChance = 10 + 5 * (GetWorldTime().GetDaysRun())
-		'iterate several times in case template with production limit and protected title is involved
-		For Local i:Int = 0 Until 20
-			filter.avoidIDs = avoidTemplateIDs
+
+		'apart from avoidTemplateIDs, templates from already available scripts should not be used again
```

Comment:
Ich wuerde den code eher folgend machen:
(ungetestet):
```
		'apart from avoidTemplateIDs, templates from already available scripts should not be used again
		Local availableScriptList:TList = GetAvailableScriptList()
		'presize array in advance instead of adding a new one-entry-array on each for-loop
		Local alreadyAvailable:Int[] = New Int[availableScriptList.Count()]
		Local scIndex:Int = 0
		for local sc:TScript = EachIn GetAvailableScriptList()
			alreadyAvailable[scIndex] = sc.basedOnScriptTemplateID
			scIndex :+ 1
		Next
		'theoretically list.count() could be higher than actual TScript count (it could contain incompatible objects..)
		If scIndex < alreadyAvailable.length Then alreadyAvailable = alreadyAvailable[..scIndex]
```
Dein Ansatz funktioniert und ist sicher noch dutzendfach im Code so aehnlich zu finden. Allerdings erzeugst Du da fuer jeden Eintrag einen eigenen Array, der wird dann mit dem bestehenden Array zu einem _neuen_ Array zusammengefuehrt. bei 20 Eintraegen sind das also 20 _neue_ Arrays plus 20 neue (groessere) Arrays die als "alreadyAvailable" hinterlegt werden. Der obige Ansatz kommt mit 1 Array + 1 moeglichem "angepasstem" Array aus.

Fuer's Prototyping ist dein Ansatz aber auch mein favorisierter.

Reply:
Wenn die Liste inkompatible Typen enthalten könnte, müsste man dann nicht noch prüfen, das sc auch wirklich nicht null ist?

Reply:
Nein, da der enumerator sicherstellt, dass eben nur kompatible typen/instanzen in "sc" landen.

Sprich, das in sc landet, ist immer "nicht null, und ein gueltiges TScript"